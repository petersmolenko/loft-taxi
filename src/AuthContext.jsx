import React, { useState } from "react";
import { Grid, Typography } from "@material-ui/core";

const AuthContext = React.createContext(null);

const AuthProvider = (props) => {
    const [isLoggedIn, setIsLoggedIn] = useState(false);
    const login = (email, password) => {
        setIsLoggedIn(true);
        console.log(`User ${email} is authorized.`);
    };
    const logout = () => setIsLoggedIn(false);
    const { children } = props;

    return (
        <AuthContext.Provider value={{ isLoggedIn, login, logout }}>
            {children}
        </AuthContext.Provider>
    );
};

const AuthHOC = (WrappedComponent) => (props) => (
    <AuthContext.Consumer>
        {(contextProps) => (
            <>
                {contextProps.isLoggedIn ? (
                    <WrappedComponent {...contextProps} {...props} />
                ) : (
                    <Grid
                        item
                        container
                        justify="center"
                        alignItems="center"
                        style={{ flexGrow: 1 }}
                    >
                        <Typography
                            variant="h3"
                            component="h3"
                            style={{ color: "#ffc617" }}
                        >
                            The user is not logged in:(
                        </Typography>
                    </Grid>
                )}
            </>
        )}
    </AuthContext.Consumer>
);

export { AuthContext, AuthProvider, AuthHOC };
