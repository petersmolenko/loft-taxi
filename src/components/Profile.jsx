import React from "react";
import { Grid, Paper, Typography, TextField, Button } from "@material-ui/core";
import cardLogo from "../assets/mc_symbol.svg";
import PropTypes from "prop-types";
import { AuthHOC } from "../AuthContext.jsx";

const Profile = ({ stopSubmit }) => (
    <Grid item xs={11} md={9} lg={7}>
        <Paper elevation={0} className="AppForm">
            <Typography variant="h4" component="h4">
                Профиль
            </Typography>
            <Typography
                variant="subtitle1"
                component="p"
                className="AppForm_margin"
                style={{ color: "rgba(0, 0, 0, 0.54)" }}
            >
                Способ оплаты
            </Typography>
            <form noValidate autoComplete="off">
                <Grid container justify="center" direction="column">
                    <Grid
                        item
                        container
                        justify="space-between"
                        spacing={3}
                        className="AppForm_margin"
                    >
                        <Grid item xs={6}>
                            <Paper
                                elevation={3}
                                style={{ padding: "2.5rem 2rem" }}
                            >
                                <div style={{ position: "relative" }}>
                                    <img
                                        src={cardLogo}
                                        style={{
                                            position: "absolute",
                                            right: 0,
                                            top: "-.5rem",
                                            width: "2rem",
                                        }}
                                        alt="card logo"
                                    />
                                </div>
                                <TextField
                                    required
                                    id="userCardNum"
                                    label="Номер карты"
                                    className="AppForm_margin"
                                    fullWidth
                                />
                                <TextField
                                    required
                                    id="userCardDuration"
                                    label="Срок действия"
                                    fullWidth
                                />
                            </Paper>
                        </Grid>
                        <Grid item xs={6}>
                            <Paper
                                elevation={3}
                                style={{ padding: "2.5rem 2rem" }}
                            >
                                <TextField
                                    required
                                    id="userCardHolder"
                                    label="Имя владельца"
                                    className="AppForm_margin"
                                    fullWidth
                                />
                                <TextField
                                    required
                                    id="userCardCVC"
                                    label="CVC"
                                    fullWidth
                                />
                            </Paper>
                        </Grid>
                    </Grid>
                    <Grid item align="center">
                        <Button
                            variant="contained"
                            color="primary"
                            style={{
                                background: "#ffc617",
                                color: "rgba(0, 0, 0, 0.87)",
                            }}
                            onClick={stopSubmit}
                        >
                            Сохранить
                        </Button>
                    </Grid>
                </Grid>
            </form>
        </Paper>
    </Grid>
);

Profile.propTypes = {
    stopSubmit: PropTypes.func
};

export default AuthHOC(Profile);
