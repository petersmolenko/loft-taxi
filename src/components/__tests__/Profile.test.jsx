import React, { useContext } from "react";
import { fireEvent, render } from "@testing-library/react";
import renderer from "react-test-renderer";
import Profile from "../Profile";
import { AuthProvider, AuthContext } from "../../AuthContext";

describe("Profile testing", () => {
    it("renders correctly (not authorized)", () => {
        const handler = jest.fn();
        const WrapperComponent = ({ isLogin, children }) => {
            const { login } = useContext(AuthContext);
            if (isLogin) login();
            return <>{children}</>;
        };
        const tree = renderer
            .create(
                <AuthProvider>
                    <WrapperComponent isLogin={false}>
                        <Profile stopSubmit={handler} />
                    </WrapperComponent>
                </AuthProvider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
    it("renders correctly (authorized)", () => {
        const handler = jest.fn();
        const WrapperComponent = ({ isLogin, children }) => {
            const { login } = useContext(AuthContext);
            if (isLogin) login();
            return <>{children}</>;
        };
        const tree = renderer
            .create(
                <AuthProvider>
                    <WrapperComponent isLogin={true}>
                        <Profile stopSubmit={handler} />
                    </WrapperComponent>
                </AuthProvider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("save-button click", () => {
        const handler = jest.fn();
        const WrapperComponent = ({ isLogin, children }) => {
            const { login } = useContext(AuthContext);
            if (isLogin) login();
            return <>{children}</>;
        };
        const { getByText } = render(
            <AuthProvider>
                <WrapperComponent isLogin={true}>
                    <Profile stopSubmit={handler} />
                </WrapperComponent>
            </AuthProvider>
        );
        fireEvent.click(getByText("Сохранить"));
        expect(handler.mock.calls.length).toBe(1);
    });
});
