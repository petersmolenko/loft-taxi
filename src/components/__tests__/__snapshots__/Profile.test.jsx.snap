// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile testing renders correctly (authorized) 1`] = `
<div
  className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-11 MuiGrid-grid-md-9 MuiGrid-grid-lg-7"
>
  <div
    className="MuiPaper-root AppForm MuiPaper-elevation0 MuiPaper-rounded"
  >
    <h4
      className="MuiTypography-root MuiTypography-h4"
    >
      Профиль
    </h4>
    <p
      className="MuiTypography-root AppForm_margin MuiTypography-subtitle1"
      style={
        Object {
          "color": "rgba(0, 0, 0, 0.54)",
        }
      }
    >
      Способ оплаты
    </p>
    <form
      autoComplete="off"
      noValidate={true}
    >
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-justify-xs-center"
      >
        <div
          className="MuiGrid-root AppForm_margin MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-item MuiGrid-justify-xs-space-between"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
          >
            <div
              className="MuiPaper-root MuiPaper-elevation3 MuiPaper-rounded"
              style={
                Object {
                  "padding": "2.5rem 2rem",
                }
              }
            >
              <div
                style={
                  Object {
                    "position": "relative",
                  }
                }
              >
                <img
                  alt="card logo"
                  src="mc_symbol.svg"
                  style={
                    Object {
                      "position": "absolute",
                      "right": 0,
                      "top": "-.5rem",
                      "width": "2rem",
                    }
                  }
                />
              </div>
              <div
                className="MuiFormControl-root MuiTextField-root AppForm_margin MuiFormControl-fullWidth"
              >
                <label
                  className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
                  data-shrink={false}
                  htmlFor="userCardNum"
                  id="userCardNum-label"
                >
                  Номер карты
                  <span
                    className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
                  onClick={[Function]}
                >
                  <input
                    aria-invalid={false}
                    autoFocus={false}
                    className="MuiInputBase-input MuiInput-input"
                    disabled={false}
                    id="userCardNum"
                    onAnimationStart={[Function]}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    required={true}
                    type="text"
                  />
                </div>
              </div>
              <div
                className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
              >
                <label
                  className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
                  data-shrink={false}
                  htmlFor="userCardDuration"
                  id="userCardDuration-label"
                >
                  Срок действия
                  <span
                    className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
                  onClick={[Function]}
                >
                  <input
                    aria-invalid={false}
                    autoFocus={false}
                    className="MuiInputBase-input MuiInput-input"
                    disabled={false}
                    id="userCardDuration"
                    onAnimationStart={[Function]}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    required={true}
                    type="text"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
          >
            <div
              className="MuiPaper-root MuiPaper-elevation3 MuiPaper-rounded"
              style={
                Object {
                  "padding": "2.5rem 2rem",
                }
              }
            >
              <div
                className="MuiFormControl-root MuiTextField-root AppForm_margin MuiFormControl-fullWidth"
              >
                <label
                  className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
                  data-shrink={false}
                  htmlFor="userCardHolder"
                  id="userCardHolder-label"
                >
                  Имя владельца
                  <span
                    className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
                  onClick={[Function]}
                >
                  <input
                    aria-invalid={false}
                    autoFocus={false}
                    className="MuiInputBase-input MuiInput-input"
                    disabled={false}
                    id="userCardHolder"
                    onAnimationStart={[Function]}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    required={true}
                    type="text"
                  />
                </div>
              </div>
              <div
                className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
              >
                <label
                  className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
                  data-shrink={false}
                  htmlFor="userCardCVC"
                  id="userCardCVC-label"
                >
                  CVC
                  <span
                    className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
                  onClick={[Function]}
                >
                  <input
                    aria-invalid={false}
                    autoFocus={false}
                    className="MuiInputBase-input MuiInput-input"
                    disabled={false}
                    id="userCardCVC"
                    onAnimationStart={[Function]}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    required={true}
                    type="text"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          align="center"
          className="MuiGrid-root MuiGrid-item"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
            disabled={false}
            onBlur={[Function]}
            onClick={[MockFunction]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "background": "#ffc617",
                "color": "rgba(0, 0, 0, 0.87)",
              }
            }
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              Сохранить
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`Profile testing renders correctly (not authorized) 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
  style={
    Object {
      "flexGrow": 1,
    }
  }
>
  <h3
    className="MuiTypography-root MuiTypography-h3"
    style={
      Object {
        "color": "#ffc617",
      }
    }
  >
    The user is not logged in:(
  </h3>
</div>
`;
